### ISSUE TITLE
Support bytes and os.PathLike and filesystem encoding using os.fsencode

### ISSUE THREAD

**Issue by graingert** [Roche/pyreadstat#123]
Describe the issue

Currently only str and PathLib objects are supported as file_path, but any PathLike, bytes or str should work. Also paths are encoded as utf8, when some paths on linux may be arbitrary bytes.

To Reproduce
Steps to reproduce the behaviour.
eg:
```python
@attr.frozen
class MyPath:
    _path: str

    def __fspath__(self):
        return self._path

read_sav(MyPath("example"))
```

**Comment by ofajardo** [comment]
thanks for the suggestion. I am unfamiliar with PathLike objects, therefore a couple of questions:

* I guess the method__fspath__ is the one that defines an object as PathLike, therefore I have to check for getattr(MyPath, "__fspath__") to check if the user used a PathLike object?
* What is the conventional way to get the PathLike object as a string? just str(MyPath)?
* How do you actually use it, for instance when you say: 

```
read_sav(MyPath("example"))
```

what is example actually? a full path like /path/to/wherever/file.sav would do?
(this last question is in order to write the unittest)

**Comment by graingert** [comment]
You just call `filepath_bytes = os.fsencode(filepath)`

**Comment by ofajardo** [comment]
@graingert , is it good?

**Comment by graingert** [comment]
lgtm, remember os.fsencode is py3.3+

**Comment by ofajardo** [comment]
ok, that's good to take into consideration, I'll add something in order not to break python 2.7.

**Comment by graingert** [comment]
You can use
https://docs.python.org/3/library/sys.html#sys.getfilesystemencoding but
it's slightly broken on py2.7 windows

**Comment by ofajardo** [comment]
I was thinking on being a bit lazy about python 2.7 and assume that the str in 'utf8'. I did this: 

https://github.com/Roche/pyreadstat/commit/3d7a520786158597319975515d5cad984ed4e2b8

But now I am remembering that in pythn 2.7 str is already bytes, so the other option was unicode?

**Comment by ofajardo** [comment]
OK, hope this is better https://github.com/Roche/pyreadstat/commit/06087c3fa39183cd4e017c2a12200e4e0626617a let me know if you have any comment

**Comment by ofajardo** [comment]
@graingert I am building in all platforms and python versions and it is working fine everywhere, except in Linux with python 3.6 (while windows and Mac Python 3.6 is fine, all other versions are fine).  The test that is failing is when trying to read [this file](https://github.com/Roche/pyreadstat/tree/master/test_data/%C3%ADntern%C3%A1tion%C3%A1l), where the folder has a non ascii-character. Do you have any suggestion?

```
======================================================================
ERROR: test_sas7bdat_international (__main__.TestBasic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../pyreadstat/tests/test_basic.py", line 202, in test_sas7bdat_international
    df, meta = pyreadstat.read_sas7bdat(os.path.join(self.international_data_folder, "sample.sas7bdat"))
  File "pyreadstat/pyreadstat.pyx", line 116, in pyreadstat.pyreadstat.read_sas7bdat
  File "pyreadstat/_readstat_parser.pyx", line 983, in pyreadstat._readstat_parser.run_conversion
  File "/usr/lib/python3.6/os.py", line 802, in fsencode
    return filename.encode(encoding, errors)
UnicodeEncodeError: 'ascii' codec can't encode character '\xed' in position 25: ordinal not in range(128)
```

**Comment by graingert** [comment]
On Linux there's not really a "filesystem encoding" because filesystem paths are opache bytes and so python has to emulate a "filesystem encoding" by guessing from the locale.

So strictly speaking this is the correct behaviour for pyreadstat. You passed a text pathname that cannot be ASCII encoded and your environment claims that text is encoded using ASCII and so this situation cannot be resolved. I'd expect `open(` to fail in the same way.

You can work around this by:

* fixing the locale to match the encoding you want to use
* passing a bytes filename on platforms that only support bytes filenames
* catch this error and call `os.fsdecode(filepath).decode("utf8", "surrogateescape")` (and probably raise a DeprecationWarning

**Comment by ofajardo** [comment]
Ok, following your last suggestion, this fixes the issue in Python 3.6.

```
    if hasattr(os, 'fsencode'):
        try:
            filename_bytes = os.fsencode(filename_path)
        except UnicodeError:
            filename_bytes = str(filename_path).encode('utf-8', 'surrogateescape')
```

**Comment by graingert** [comment]
I think you want:

```python
     if hasattr(os, 'fsencode'):
         try:
             filename_bytes = os.fsencode(filename_path)
         except UnicodeEncodeError:
             filename_bytes = os.fsdecode(filename_path).encode('utf-8', 'surrogateescape')
```

**Comment by ofajardo** [comment]
Yeah, I tried, but that also raises an exception (same message about ascii encoding failing)

**Comment by graingert** [comment]
> Yeah, I tried, but that also raises an exception (same message about ascii encoding failing)

Hmm, show the error?

**Comment by ofajardo** [comment]
here it is:

```
(Pdb) mypath = os.path.join(self.international_data_folder, "sample.sas7bdat")
(Pdb) os.fsdecode(mypath)
*** UnicodeEncodeError: 'ascii' codec can't encode character '\xed' in position 45: ordinal not in range(128)
```

**Comment by graingert** [comment]
Hmm that's weird, os.fsdecode is documented as " return str unchanged."

**Comment by ofajardo** [comment]
Maybe a bug in 3.6?

**Comment by ofajardo** [comment]
well, no idea. On the same ubuntu machine python 3.6 says os.getfilesystemencoding is ascii and in python 3.7 it says it is utf-8. I wanted to change the filesystemencoding for 3.7 to see if it would do the same or not with os.fsdecode, but I can't get it to work chaning LC_ALL, or LANG (setting it to C).

**Comment by graingert** [comment]
```
Python 3.6.13 (default, Feb 20 2021, 21:40:46) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale, os
>>> locale.getpreferredencoding()
'ANSI_X3.4-1968'
>>> os.fsencode(u"\u2603")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/os.py", line 802, in fsencode
    return filename.encode(encoding, errors)
UnicodeEncodeError: 'ascii' codec can't encode character '\u2603' in position 0: ordinal not in range(128)
>>> os.fsdecode(u"\u2603").encode("utf8", "surrogateescape")
b'\xe2\x98\x83'
>>> 
```

**Comment by graingert** [comment]
here's the code, it passes `str` through unchanged:
https://github.com/python/cpython/blob/5b1e50256b6532667b6d31debc350f6c7d3f30aa/Lib/os.py#L816-L817

**Comment by graingert** [comment]
@ofajardo try:

```python
mypath = os.path.join(self.international_data_folder, "sample.sas7bdat")
os.fspath(mypath)
```

**Comment by graingert** [comment]
it might be that it's failing in pdb when it tries to print out the result

**Comment by ofajardo** [comment]
os.fspath in (my?) python 3.6 doesn't like strings: (nothing to do with pdb actually)

```
filename_bytes = os.fspath(filename_path)
TypeError: Expected bytes, got str
```

But this other suggestion works well : 

```
filename_bytes = os.fsdecode(filename_path).encode("utf-8", "surrogateencoding")
```

so now, I am wondering if I should do that for any python function, or do os.fsencode(filename_path) by default, and only if UnicodeError is raised then use os.fsdecode etc. I guess the second? Sorry, all of this stuff confuses me.

**Comment by graingert** [comment]
Doing:

`os.fsencode(filename_path)` by default, and only if UnicodeEncodeError is raised then use `os.fsdecode` makes the most sense.

I think you should also warn a deprecation warning when falling back to utf8. Users should configure their locale correctly or pass bytes

**Comment by graingert** [comment]
the error you get is different to mine too:

```
Python 3.6.13 (default, Feb 20 2021, 21:40:46) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.fspath(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected str, bytes or os.PathLike object, not int
```

what's your py3.6 minor version?

**Comment by ofajardo** [comment]
makes sense, I'll add that warning. I'm intrigued on how to change the locale on python 3.6. I changed the locale of the system and still python didn't care, it stayed in ascii.

python version:

```
Python 3.6.13 |Anaconda, Inc.| (default, Feb 23 2021, 21:15:04)
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
```
Another question, if fsencode is not an attribute of os I am doing this below. You said it would be good to get the filesystemencoding, but I am thinking that there could be cases as we have seen in python 3.6 where the system encoding doesn't match your strings, the user doesn't know how to change the python encoding (as myself right now with python 3.6) and then I was simply forcing the user to use utf-8 (as it was until now, so let's say legacy mode ... I can raise a warning about that). Any opinion? I guess also these cases will tend to disappear.

```
    else:
        IF PY_MAJOR_VERSION >2: # Python 3.2 or earlier
            if type(filename_path) == str:
                filename_bytes = filename_path.encode('utf-8')
            elif type(filename_path) == bytes:
                filename_bytes = filename_path
            else:
                raise PyreadstatError("path must be either str or bytes")
        ELSE: # Python 2
            if type(filename_path) not in (str, bytes, unicode):
                raise PyreadstatError("path must be str, bytes or unicode")
            filename_bytes = filename_path.encode('utf-8')
```

**Comment by graingert** [comment]
> but I am thinking that there could be cases as we have seen in python 3.6 where the system encoding doesn't match your strings

str has no associated encoding, it's just a sequence of Unicode code points

**Comment by ofajardo** [comment]
in python 2 yes, I guess (the .encode is for unicode), but in Python <=3.2 I guess str have encoding ? (I jumped from 2.7 to 3.6 directly, so no idea how 3.2 and earlier behaves)

**Comment by ofajardo** [comment]
the feature is now available in version 1.1.1
